homeassistant/components/synology_dsm/sensor.py

```diff
@@ -0,0 +1,258 @@
+"""Support for Synology DSM Sensors."""
+from datetime import timedelta
+from typing import Dict
+
+from synology_dsm import SynologyDSM
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    ATTR_ATTRIBUTION,
+    CONF_API_VERSION,
+    CONF_DISKS,
+    CONF_HOST,
+    CONF_NAME,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_SSL,
+    CONF_USERNAME,
+    DATA_MEGABYTES,
+    DATA_RATE_KILOBYTES_PER_SECOND,
+    TEMP_CELSIUS,
+)
+from homeassistant.helpers.dispatcher import async_dispatcher_connect, dispatcher_send
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import HomeAssistantType
+
+from .const import (
+    CONF_VOLUMES,
+    DEFAULT_DSM_VERSION,
+    DOMAIN,
+    SERVICE_UPDATE,
+    STORAGE_DISK_SENSORS,
+    STORAGE_VOL_SENSORS,
+    TEMP_SENSORS_KEYS,
+    UTILISATION_SENSORS,
+)
+
+ATTRIBUTION = "Data provided by Synology"
+
+SCAN_INTERVAL = timedelta(minutes=15)
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up the Synology NAS Sensor."""
+    name = entry.data[CONF_NAME]
+    host = entry.data[CONF_HOST]
+    port = entry.data[CONF_PORT]
+    username = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+    unit = hass.config.units.temperature_unit
+    use_ssl = entry.data[CONF_SSL]
+    api_version = entry.data.get(CONF_API_VERSION, DEFAULT_DSM_VERSION)
+
+    api = SynoApi(hass, host, port, username, password, unit, use_ssl, api_version)
+
+    await hass.async_add_executor_job(api.update)
+
```

Comment:
Does it make sense top fetch all the stats here instead? If I change out a disk now I'll have to re-setup the config.

```

        information = await self.hass.async_add_executor_job(
            getattr, api, "information"
        )
        utilisation = await self.hass.async_add_executor_job(
            getattr, api, "utilisation"
        )
        storage = await self.hass.async_add_executor_job(getattr, api, "storage")
```

Reply:
Probably would be better to now spawn 4 jobs and do it all in one as well.  I've been told not to do that in the past
```
 syno_inventory =  await self.hass.async_add_executor_job(_get_device_inventory(hass,api))
``
```
def _get_device_inventory(hass, api):
 api.update()
 return {
   SYNO_API_INFORMATION: api.information()
   SYNO_API_UTILISATION: api.utilisation()
   SYNO_API_STORAGE: api.storage()
 }
```

Reply:
> Does it make sense top fetch all the stats here instead? If I change out a disk now I'll have to re-setup the config.

As before, just restart HA.

----

> Probably would be better to not spawn 4 jobs and do it all in one as well. I've been told not to do that in the past

As I will add an Option flow or something to param set specific disk or volume, I don't think it's worth it.

Reply:
* Will be fixed in a followup with options flow